### PR TITLE
Functional: itir transform improvements

### PR DESCRIPTION
**Pull request by tehrengruber**
- Add option to `InlineLambdas` preserving opcount
- ~Add option to `InlineLifts` pass perserving number of shifts~
- Add common subexpression elimination again (#754)
- Repeated elimination of lifts and lambda calls
- Remove unnecessary lifts for use_tmps option
- Improve common subexpression elimination pass to avoid collecting child subexpressions of eliminated expressions

More ellaborate description follows.

Examples from FVM Advection sphere solver:

Before:
```
λ(
  rho,
  dt,
  vol,
  gac,
  vel_x,
  vel_y,
  dual_face_orientation,
  dual_face_normal_weighted_x,
  dual_face_normal_weighted_y
) → 
  ·(λ(flux__0) → 
      (↑(λ(rho, dt, vol, gac, flux, dual_face_orientation) → 
           ·rho
           - ·dt / (·vol × ·gac)
             × reduce(λ(acc, flux__0, dual_face_orientation__1) → acc + flux__0 × dual_face_orientation__1, 0)(
                 ⟪V2Eₒ⟫(flux), dual_face_orientation
               )))(rho, dt, vol, gac, flux__0, dual_face_orientation))(
     (↑(λ(rho, vel_x, vel_y, dual_face_normal_weighted_x, dual_face_normal_weighted_y) → 
          ·(λ(wnv__0) → 
              (↑(λ(wnv__0, rho) → if ·wnv__0 > 0.0 then ·⟪E2Vₒ, 0⟫(rho) × ·wnv__0 else ·⟪E2Vₒ, 1⟫(rho) × ·wnv__0))(
                wnv__0, rho
              ))((↑(λ(vel_x, dual_face_normal_weighted_x, vel_y, dual_face_normal_weighted_y) → 
                      ·vel_x × ·dual_face_normal_weighted_x + ·vel_y × ·dual_face_normal_weighted_y))(
                   vel_x, dual_face_normal_weighted_x, vel_y, dual_face_normal_weighted_y
                 ))))(rho, vel_x, vel_y, dual_face_normal_weighted_x, dual_face_normal_weighted_y)
   )
```

After:
```
λ(
  rho,
  dt,
  vol,
  gac,
  vel_x,
  vel_y,
  dual_face_orientation,
  dual_face_normal_weighted_x,
  dual_face_normal_weighted_y
) → 
  ·rho
  - ·dt / (·vol × ·gac)
    × reduce(λ(acc, flux__0_, dual_face_orientation__1) → acc + flux__0_ × dual_face_orientation__1, 0)(
        ⟪V2Eₒ⟫((↑(λ(rho, vel_x, vel_y, dual_face_normal_weighted_x, dual_face_normal_weighted_y) → 
                    (λ(_cs_1) → if _cs_1 > 0.0 then ·⟪E2Vₒ, 0⟫(rho) × _cs_1 else ·⟪E2Vₒ, 1⟫(rho) × _cs_1)(
                      ·vel_x × ·dual_face_normal_weighted_x + ·vel_y × ·dual_face_normal_weighted_y
                    )))(rho, vel_x, vel_y, dual_face_normal_weighted_x, dual_face_normal_weighted_y)),
        dual_face_orientation
      )
```

With `unroll_reduce`:

```
λ(
  rho,
  dt,
  vol,
  gac,
  vel_x,
  vel_y,
  dual_face_orientation,
  dual_face_normal_weighted_x,
  dual_face_normal_weighted_y
) → 
  ·rho
  - ·dt / (·vol × ·gac)
    × (λ(_step_3) → _step_3(_step_3(_step_3(_step_3(_step_3(_step_3(0, 0ₒ), 1ₒ), 2ₒ), 3ₒ), 4ₒ), 5ₒ))(
        λ(_acc_1, _i_2) → 
          (λ(_cs_5, _cs_6, _cs_7, _cs_8, _cs_9) → 
             if   can_deref(_cs_5) ∧ can_deref(_cs_6) ∧ can_deref(_cs_7) ∧ can_deref(_cs_8) ∧ can_deref(_cs_9)
             then _acc_1 + (λ(_cs_4) → if _cs_4 > 0.0 then ·⟪E2Vₒ, 0⟫(_cs_5) × _cs_4 else ·⟪E2Vₒ, 1⟫(_cs_5) × _cs_4)(
                             ·_cs_6 × ·_cs_8 + ·_cs_7 × ·_cs_9
                           ) × ·⟪_i_2⟫(dual_face_orientation)
             else _acc_1)(⟪V2Eₒ, _i_2⟫(rho),
                          ⟪V2Eₒ, _i_2⟫(vel_x),
                          ⟪V2Eₒ, _i_2⟫(vel_y),
                          ⟪V2Eₒ, _i_2⟫(dual_face_normal_weighted_x),
                          ⟪V2Eₒ, _i_2⟫(dual_face_normal_weighted_y))
      )
```

### CODE REVIEW

**Review by fthaler — CHANGES_REQUESTED** [review]
Looks good, but I’d like to reduce the conflicts with scan-canonicalization a bit, namely exclude the changes to InlineLifts.

File: src/functional/iterator/embedded.py [hunk 1]
```diff
@@ -567,8 +567,9 @@ def deref(self) -> Any:
         shifted_pos = self.pos.copy()
         axes = _get_axes(self.field)
 
-        if not all(axis.value in shifted_pos.keys() for axis in axes if axis is not None):
-            raise IndexError("Iterator position doesn't point to valid location for its field.")
+        if __debug__:
+            if not all(axis.value in shifted_pos.keys() for axis in axes if axis is not None):
+                raise IndexError("Iterator position doesn't point to valid location for its field.")
```
Comment (havogt) [inline src/functional/iterator/embedded.py:572]: What's this about? Performance?

File: src/functional/iterator/transforms/pass_manager.py [hunk 2]
```diff
@@ -7,34 +9,45 @@
...
-        ir = InlineLifts().visit(ir)
-    ir = InlineLambdas().visit(ir)
+    for _ in range(10):
+        inlined_ir = InlineLifts.apply(ir, preserve_shift_count=use_tmps)
+        inlined_ir = InlineLambdas.apply(inlined_ir)
+        if inlined_ir == ir:
+            break
+        ir = inlined_ir
+    else:
+        raise RuntimeError("Inlining lift and lambdas did not converge.")
```
Comment (havogt) [inline src/functional/iterator/transforms/pass_manager.py:29]: Maybe we can merge this separately in case it fixes the problem in Sam's stencil. I'll try it.

Comment (havogt) [reply src/functional/iterator/transforms/pass_manager.py:29]: It does fix the problem. I'll make a PR.

File: src/functional/iterator/transforms/inline_lifts.py [hunk 3]
```diff
@@ -1,40 +1,49 @@
...
+@dataclass
 class InlineLifts(NodeTranslator):
-    @staticmethod
-    def _is_lift(node: ir.Node):
-        return (
-            isinstance(node, ir.FunCall)
-            and isinstance(node.fun, ir.FunCall)
-            and node.fun.fun == ir.SymRef(id="lift")
-        )
+    preserve_shift_count: bool
```
[the remark this answers is not in the document]
Comment (tehrengruber) [reply src/functional/iterator/transforms/inline_lifts.py]: Moved that part to https://github.com/GridTools/gt4py/pull/886.

File: src/functional/iterator/embedded.py (hunk 1 above)
Comment (tehrengruber) [reply src/functional/iterator/embedded.py:572]: Indeed. It doesn't give that much though so I could remove it.

File: src/functional/iterator/transforms/pass_manager.py [hunk 4]
```diff
@@ -7,41 +9,49 @@
...
     if unroll_reduce:
         for _ in range(10):
             unrolled = UnrollReduce().visit(ir, offset_provider=offset_provider)
             if unrolled == ir:
                 break
             ir = unrolled
             ir = NormalizeShifts().visit(ir)
-            if not use_tmps:
-                ir = InlineLifts().visit(ir)
+            ir = InlineLifts().visit(ir)
```
Comment (fthaler) [inline src/functional/iterator/transforms/pass_manager.py:43]:
```suggestion
            if not use_tmps:
                ir = InlineLifts().visit(ir)
```
Why did you remove the check? If it’s not there, the `CreateGlobalTmps` pass below will always be a no-op.

**Review by fthaler — CHANGES_REQUESTED** [review]

File: src/functional/iterator/transforms/pass_manager.py (hunk 4 above)
Comment (tehrengruber) [reply src/functional/iterator/transforms/pass_manager.py]: Mh, probably a misstake during merging. Fixed.

**Review by fthaler — APPROVED** [review]
